### PR TITLE
[WIP] Opening files in Viewer using URL-string parameter.

### PR DESCRIPTION
**Pull request by kegor**
**Related PR:**
PR with similar feature but using the provided UI on Angular: https://github.com/groupdocs-total/GroupDocs.Total-Angular/pull/200

**Screenshots:**
_Result of the remote request for file content using relative path:_
image

_Result of the remote request for file content using URL:_
image

### CODE REVIEW

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476&bid=19532978)

File: src/Products/Viewer/Controllers/ViewerApiController.cs [hunk 1]
```diff
@@ -674,6 +674,7 @@ private static LoadDocumentEntity GetDocumentPages(PostedDataEntity postedData,
             // get/set parameters
             string documentGuid = postedData.guid;
             string password = string.IsNullOrEmpty(postedData.password) ? null : postedData.password;
+            if (!File.Exists(documentGuid)) { throw new Exception("File not found."); }
```
Comment (bobkovalex) [inline src/Products/Viewer/Controllers/ViewerApiController.cs:677]: Codacy found an issue: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476)

Comment (egorovpavel) [inline src/Products/Viewer/Controllers/ViewerApiController.cs:677]: Codacy found an issue: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475&bid=19532977)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475&bid=19532977)

File: src/Products/Viewer/Controllers/ViewerApiController.cs [hunk 2]
```diff
@@ -674,6 +674,10 @@ private static LoadDocumentEntity GetDocumentPages(PostedDataEntity postedData,
             // get/set parameters
             string documentGuid = postedData.guid;
             string password = string.IsNullOrEmpty(postedData.password) ? null : postedData.password;
+            if (!File.Exists(documentGuid)) 
+            { 
+                throw new Exception("File not found.");
```
Comment (egorovpavel) [inline src/Products/Viewer/Controllers/ViewerApiController.cs:679]: Codacy found an issue: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476&bid=19532978)

File: src/Products/Viewer/Controllers/ViewerApiController.cs (hunk 2 above)
Comment (bobkovalex) [inline src/Products/Viewer/Controllers/ViewerApiController.cs:679]: Codacy found an issue: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476)

File: src/Products/Viewer/Util/DocumentDataLoader.cs [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Web;
+using System.Xml.Linq;
+
+namespace GroupDocs.Viewer.MVC.Products.Viewer.Util
+{
+    public class DocumentDataLoader
```
Comment (egorovpavel) [inline src/Products/Viewer/Util/DocumentDataLoader.cs:20]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 12
           

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  13
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475&bid=19532977)

File: src/Products/Common/Entity/Web/PageDescriptionEntity.cs [hunk 4]
```diff
@@ -31,7 +31,7 @@ public class PageDescriptionEntity
         /// Page content data.
         /// </summary>
         [JsonProperty]
-        private string data;
+        public string data;
```
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/PageDescriptionEntity.cs:34]: Codacy found an issue: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Products/Common/Entity/Web/LoadDocumentEntity.cs [hunk 5]
```diff
@@ -12,25 +12,25 @@ public class LoadDocumentEntity
...
         [JsonProperty]
-        private List<PageDescriptionEntity> pages = new List<PageDescriptionEntity>();
+        public List<PageDescriptionEntity> pages = new List<PageDescriptionEntity>();
 
         /// <summary>
         /// Document print allowed flag.
         /// </summary>
         [JsonProperty]
-        private bool printAllowed = true;
+        public bool printAllowed = true;
```
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:27]: Codacy found an issue: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:27]: Codacy found an issue: [Make 'printAllowed' private.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Products/Common/Entity/Web/LoadDocumentEntity.cs [hunk 6]
```diff
@@ -12,25 +12,25 @@ public class LoadDocumentEntity
...
 
         /// <summary>
         /// Document show grid lines flag (for Excel files). 
         /// </summary>
         [JsonProperty]
-        private bool showGridLines = true;
+        public bool showGridLines = true;
```
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:33]: Codacy found an issue: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Products/Common/Entity/Web/LoadDocumentEntity.cs [hunk 7]
```diff
@@ -12,25 +12,25 @@ public class LoadDocumentEntity
         /// Document absolute path.
         /// </summary>
         [JsonProperty]
-        private string guid;
+        public string guid;
```
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:15]: Codacy found an issue: [Make 'guid' private.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Products/Common/Entity/Web/LoadDocumentEntity.cs [hunk 8]
```diff
@@ -12,25 +12,25 @@ public class LoadDocumentEntity
...
 
         /// <summary>
         /// Collection of the document pages with their data.
         /// </summary>
         [JsonProperty]
-        private List<PageDescriptionEntity> pages = new List<PageDescriptionEntity>();
+        public List<PageDescriptionEntity> pages = new List<PageDescriptionEntity>();
```
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:21]: Codacy found an issue: [Make 'pages' private.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Products/Common/Entity/Web/LoadDocumentEntity.cs (hunk 7 above)
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:15]: Codacy found an issue: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Products/Common/Entity/Web/LoadDocumentEntity.cs (hunk 8 above)
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:21]: Codacy found an issue: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Products/Common/Entity/Web/LoadDocumentEntity.cs (hunk 6 above)
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/LoadDocumentEntity.cs:33]: Codacy found an issue: [Make 'showGridLines' private.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

File: src/Controllers/ViewerController.cs [hunk 9]
```diff
@@ -7,9 +10,34 @@ namespace GroupDocs.Viewer.MVC.Controllers
...
+                bool result = Uri.TryCreate(file, UriKind.Absolute, out uriResult)
+                    && (uriResult.Scheme == Uri.UriSchemeHttp || uriResult.Scheme == Uri.UriSchemeHttps);
+
+                if (result)
+                {
+                    string fileSavePath = DocumentDataLoader.GetFileSavePath(file);
+                    documentGuid = fileSavePath;
+                }
+
+                var postedData = new PostedDataEntity()
```
Comment (egorovpavel) [inline src/Controllers/ViewerController.cs:29]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 12
           

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  13
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476&bid=19532978)

File: src/Products/Common/Entity/Web/PageDescriptionEntity.cs (hunk 4 above)
Comment (egorovpavel) [inline src/Products/Common/Entity/Web/PageDescriptionEntity.cs:34]: Codacy found an issue: [Make 'data' private.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  13
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475&bid=19532977)

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  13
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476&bid=19532978)

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  13
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476&bid=19532978)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  13
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475&bid=19532977)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  15
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958475&bid=19532977)

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Util/DocumentDataLoader.cs  15
         

Clones added
============
- src/Products/Viewer/Util/DocumentDataLoader.cs  1
         

Clones removed
==============
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5958476&bid=19532978)

**Pull request merged**